Performance.md

```diff
@@ -0,0 +1,290 @@
+
+These two options can often hide misconfiguration and conflicts in `.d.ts` files, so we suggest using them sparingly.
+
+# Configuring Other Build Tools
+
+TypeScript compilation is often performed with other build tools in mind - especially when writing web apps that might involve a bundler.
+While we can only make suggestions for a few build tools, ideally these techniques can be generalized.
+
+## Concurrent Type-Checking
+
+Type-checking typically requires semantic information from other files, and can be relatively expensive compared to other steps like emitting.
+Because type-checking can take a little bit longer, it can impact the inner development loop - in other words, you might experience a longer edit/compile/retry cycle, and this might be frustrating.
+
+For this reason, some build tools can run type-checking in a concurrent process without blocking emit.
+While this means that invalid code can run before TypeScript reports an error in the tool, you'll often see errors in your editor first, and you won't be blocked for as long from running working code.
+
+An example of this in action is Webpack's `fork-ts-checker` plugin.
+
+## Isolated File Emit
+
+By default, TypeScript's emit requires semantic information that might not be local to a file.
+This is to understand how to emit features like `const enum`s and `namespace`s.
+But needing to check other files to emit of an arbitrary file can make emit slower.
+
+The need for features that need non-local information is somewhat rare - regular `enum`s can be used in place of `const enum`s, and modules can be used instead of `namespace`s.
+For that reason, TypeScript provides the `isolatedModules` flag to warn when using them.
+Using `isolatedModules` means that your codebase is safe for tools to use TypeScript APIs like `transpileModule` or alternative compilers like Babel.
+
+As an example of this in action, ts-loader provides
+
+* the `transpileOnly` flag to use `transpileModule`.
+* the `useBabel` flag
+
+<!-- TODO -->
+
+# Investigating Issues
+
+There are certain ways to get hints of what might be going wrong.
+
+## `extendedDiagnostics`
+
+You can run TypeScript with `--extendedDiagnostics` to get a printout of where the compiler is spending its time.
+
+```
+Files:                         6
+Lines:                     24906
+Nodes:                    112200
+Identifiers:               41097
+Symbols:                   27972
+Types:                      8298
+Memory used:              77984K
+Assignability cache size:  33123
+Identity cache size:           2
+Subtype cache size:            0
+I/O Read time:             0.01s
+Parse time:                0.44s
+Program time:              0.45s
+Bind time:                 0.21s
+Check time:                1.07s
+transformTime time:        0.01s
+commentTime time:          0.00s
+I/O Write time:            0.00s
+printTime time:            0.01s
+Emit time:                 0.01s
+Total time:                1.75s
```

Comment:
Worth noting that this is not actually the elapsed time?

Reply:
Specifically, I was super-confused when I saw that it wasn't the total of the preceding lines.

Reply:
Is it lower, or higher? What is the implication here?

Reply:
There are many line items, but `Total time` is the sum of only four of them: `programTime + bindTime + checkTime + emitTime`.  As such, the total will generally be less than the sum of the line items above it and less than the wall-clock time.